### ISSUE TITLE
AspNetCore authorization scopes policies helper

### ISSUE THREAD

**Issue by Tratcher** [dotnet/aspnetcore#10447]
_From @baywet on Friday, 17 May 2019 14:53:25_

**Is your feature request related to a problem? Please describe.**
If you're building an aspnetcore mvc application [Microsoft.AspNetCore.Authentication.AzureAD.UI](https://github.com/aspnet/AspNetCore/tree/master/src/Azure/AzureAD) does a fantastic job at seemlessly integrating your application with AAD from an authN standpoint.

Now let's assume you also want to build an API, and that API will have different surfaces that are secured using scopes. (ie: I don't want to grant access to all applications to the complete API endpoints exposed by my API).

Because AAD appends all the scopes as "scope1 scope2 scope3" in the scp claim instead of using a multi-valued claim ( scopes: ["scope1", "scope2", "scope3"]) like other authentication providers do, it's not possible to leverage existing scope policies libraries out there. (like identityserver4.accesstokenvalidation which provides ScopePolicy.Create).
It is also not possible to leverage OOTB aspnetcore APIs like a policy builder with a claim requirement.

**Describe the solution you'd like**
It'd be nice to have new package Microsoft.AspNetCore.Authorization.AzureAD, this package would basically contain two static mehods:
```CSharp
namespace Microsoft.AspNetCore.Authentication {
public static class AzureADAuthorizationBuilderExtensions {
public static AuthorizationBuilder AddAzureAD(this AuthorizationBuilder, string scheme...); // this method would add the AuthorizationHandler to the services
}
}
```
```CSharp
namespace Microsoft.AspNetCore.Authorization
{
    //
    // Summary:
    //     Helper for creating scope-related policies
    public static class ScopePolicy
    {
        //
        // Summary:
        //     Creates a policy to check for required scopes.
        //
        // Parameters:
        //   scopes:
        //     List of any required scopes. The token must contain at least one of the listed
        //     scopes.
        public static AuthorizationPolicy Create(params string[] scopes);
    }
}
```
Here is some example handler
```CSharp
public class OneOfTheScopesHandler : AuthorizationHandler<OneOfTheScopesRequirement>
    {
        private const string scopeName = "http://schemas.microsoft.com/identity/claims/scope";
        protected override Task HandleRequirementAsync(AuthorizationHandlerContext context, OneOfTheScopesRequirement requirement)
        {
            if (!context.User.HasClaim(x => x.Type.Equals(scopeName)))
            {
                context.Fail();
            }
            else
            {
                var currentScopes = context.User.Claims.Where(x => x.Type.Equals(scopeName))
                    .Select(x => x.Value)
                    .SelectMany(x => x.Split(new char[] { ' ' }, StringSplitOptions.RemoveEmptyEntries));
                if (currentScopes.Any(x => requirement.Scopes.Any(y => y.Equals(x, StringComparison.InvariantCultureIgnoreCase))))
                    context.Succeed(requirement);
                else context.Fail();
            }
            return Task.CompletedTask;
        }
    }
public class OneOfTheScopesRequirement: IAuthorizationRequirement
    {
        public OneOfTheScopesRequirement(params string[] scopes)
        {
            Scopes = scopes;
        }
        public string[] Scopes { get; set; }
    }
```

**Additional context**
I think all of that should seat under the aspnetcore repo I pointed to earlier but should be owned by the same team owning MSAL. So before starting any work I wanted to get the conversation going.

Having @Tratcher (https://github.com/Tratcher) part of the conversation could help as well.

_Copied from original issue: AzureAD/microsoft-authentication-library-for-dotnet#1158_

**Comment by Tratcher** [comment]
_From @Tratcher on Friday, 17 May 2019 15:02:13_

@haok for Authorization

**Comment by Tratcher** [comment]
_From @jmprieur on Wednesday, 22 May 2019 05:28:25_

@Tratcher : do we want to move this issue to the ASP.NET Core repo? Unless I'm missing something, it's not related to MSAL.NET ?
cc;   @baywet

**Comment by Tratcher** [comment]
_From @jmprieur on Wednesday, 22 May 2019 05:30:44_

cc: @MarkZuber and @kalyankrishna1 FYI

**Comment by Tratcher** [comment]
_From @baywet on Wednesday, 22 May 2019 12:17:29_

@jmprieur the reason I started the conversation here is because the way of AAD of representing the scopes in a token is not "standard" (although the way to represent scopes doesn't seem to be part of the original JWT spec). IMHO the end library would belong to the aspnet core repo but I thought people here might be interested in having a conversation before starting any work.

**Comment by Tratcher** [comment]
@baywet An alternate design would be to fix up the scopes claim during authentication so it was in the expected format afterwards. I expect this could be done today with a custom ClaimsAction.

**Comment by brockallen** [comment]
IdentityServer issues "scope" claim types, FWIW. 

```
{
  "nbf": 1558543824,
  "exp": 1558547424,
  "iss": "https://localhost:5001",
  "aud": [
    "https://localhost:5001/resources",
    "api.api1"
  ],
  "client_id": "spa",
  "sub": "99b81d5d-529e-4495-ab87-841e03ba3774",
  "auth_time": 1558543823,
  "idp": "local",
  "name": "alice@email",
  "email": "alice@email",
  "scope": [
    "openid",
    "id.custom",
    "api.api1"
  ],
  "amr": [
    "pwd"
  ]
}


But I think AAD uses `scp`.
```

**Comment by baywet** [comment]
Yep not only aad uses scp they also use an appended string containing all the scopes separated by space. I like your solution better, requires less work to build and then less work to use.

**Comment by kalyankrishna1** [comment]
Making sure we refer the corresponding spec. Ours for AAD V2 end point are 

**For request**
https://openid.net/specs/openid-connect-core-1_0.html#ScopeClaims

> Multiple scope values MAY be used by creating a space delimited, case sensitive list of ASCII scope values.

**For the claim in token** 

https://tools.ietf.org/html/rfc7662#section-3.3 
    scope

>       OPTIONAL.  A JSON string containing a space-separated list of
>       scopes associated with this token, in the format described in
>       Section 3.3 of OAuth 2.0 [RFC6749].

**Comment by brockallen** [comment]
> For the claim in token
> https://tools.ietf.org/html/rfc7662#section-3.3

That's not a token. That's the JSON being returned from introspection.

**Comment by kalyankrishna1** [comment]
Right - there's no spec or standard for format of scopes in access tokens, nor indeed any spec for access tokens in general.  There's a spec out for review that provides one, and Microsoft Identity Platform access tokens match that spec (https://tools.ietf.org/html/draft-ietf-oauth-access-token-jwt-00#section-3) with regards to the scope claim format (albeit not the name itself).

It makes sense that a library that wants to support multiple IDPs should be ready to interpret multiple access token formats, including Google or Facebook's opaque blobs.

**Comment by anurse** [comment]
Triage: We can look at this post-3.0. We think the AAD provider can do a conversion in a non-breaking way by **adding** a `scope` claim that is properly formatted and leaving the `scp` claim as-is. We may look at doing a sample on how to do this in your own app in the interim.

**Comment by Tratcher** [comment]
The AzureAD.UI packages are now obsolete and have been replaced by Microsoft.Identity.Web, anyone interested in pursuing this should do so over at https://github.com/AzureAD/microsoft-identity-web/.
